### DISCUSSION TITLE
Can't save as fmu file

### DISCUSSION THREAD

**Discussion by basatsu** [ThummeTo/FMIExport.jl#50] [Q&A]
I tried to execute this example and export .fmu file.
[https://github.com/ThummeTo/FMIExport.jl/blob/main/examples/FMI2/BouncingBall/src/BouncingBall.jl](https://github.com/ThummeTo/FMIExport.jl/blob/main/examples/FMI2/BouncingBall/src/BouncingBall.jl)

However, it did not work. I'm facing this error.
```
Warning: .drectve `-exclude-symbols:jl_ptls_table ' unrecognized
Warning: .drectve `-exclude-symbols:jl_small_typeof ' unrecognized
C:\Users\○○○○\.julia\packages\FMIBuild\zfhlh\src/../template/ME/header/FMU2_init.o:FMU2_init.c:(.text+0x214): undefined reference to `init_FMU'
collect2.exe: error: ld returned 1 exit status
ERROR: LoadError: failed process: 
・・・
・・・
```

It seems like it happend at these lines.
```
saveFMU(fmu, fmu_save_path; debug=true, compress=false) 
```

I'm not sure why this happens and "saveFMU" doesn't work.
How do I solve this?
It'd be nice if you could give me why this error happens and your solution.

Also, I discovered other similar comments here.  I'm guessing these comments says the similar errors to mine.
[https://discourse.julialang.org/t/issue-with-saving-fmu-file-in-fmiexport-jl-cannot-find-a-package-where-this-file-is-stored/120856](https://discourse.julialang.org/t/issue-with-saving-fmu-file-in-fmiexport-jl-cannot-find-a-package-where-this-file-is-stored/120856)

[https://discourse.julialang.org/t/issue-with-fmu-export-after-manipulating-model-in-julia/121084](https://discourse.julialang.org/t/issue-with-fmu-export-after-manipulating-model-in-julia/121084)

Best,

**Comment by ThummeTo** [comment]
The error you provided looks different compared to the linked posts (they are about the fact that the file containing `saveFMU` are not within a Julia-project).

Can you provide a complete example? Or explain all steps you are doing?

**Comment by basatsu** [comment]
Hi @ThummeTo,
 
Got it, here is what I did.

First, I copied this example code "BouncingBall" to my Julia file.
[https://github.com/ThummeTo/FMIExport.jl/blob/main/examples/FMI2/BouncingBall/src/BouncingBall.jl](https://github.com/ThummeTo/FMIExport.jl/blob/main/examples/FMI2/BouncingBall/src/BouncingBall.jl)

Second, I changed just 
`saveFMU(fmu, fmu_save_path; debug=true, compress=false) `
to
`saveFMU(fmu, fmu_save_path; compress=false) `

I just deleted "debug=true".
Then I executed mu Julia file in my Julia project.
However, this error happened as below. I pasted it.

```
[ Info: PackageCompiler: Done
⠙ [00m:35s] PackageCompiler: compiling incremental system imagecould not load library ""
The parameter is incorrect.
Stacktrace:
⣄ [00m:36s] PackageCompiler: compiling incremental system imagezmq_version(major_::Base.RefValue{Int32}, minor_::Base.RefValue{Int32}, patch_::Base.RefValue{Int32})
・・・
・・・
in expression starting at C:\Users\Owner\.julia\packages\ZMQ\7UMJa\src\ZMQ.jl:3
in expression starting at C:\Users\Owner\AppData\Local\Temp\jl_ZK5OjCZEdi:31
✖ [00m:36s] PackageCompiler: compiling incremental system image
ERROR: LoadError: failed process: Process(`'C:\Users\Owner\.julia\juliaup\julia-1.10.4+0.x64.w64.mingw32\bin\julia.exe' --color=yes --startup-file=no --pkgimages=no '--cpu-target=generic;sandybridge,-xsaveopt,clone_all;haswell,-rdrnd,base(1)' '--sysimage=C:\Users\Owner\AppData\Local\Temp\jl_j9dXxP\tmp_sys.dll' '--project=C:\Users\Owner\AppData\Local\Temp\fmibuildjl_mF2Fs4\merged_BouncingBall' '--output-o=C:\Users\Owner\AppData\Local\Temp\jl_diPqdklm5m.o' 'C:\Users\Owner\AppData\Local\Temp\jl_ZK5OjCZEdi'`, ProcessExited(1)) [1]

Stacktrace:
  [1] pipeline_error
    @ .\process.jl:565 [inlined]
・・・
・・・
 [13] saveFMU(fmu::FMIBase.FMU2, fmu_path::String, fmu_src_file::Nothing; standalone::Bool, compress::Bool, cleanup::Bool, removeLibDependency::Bool, removeNoExportBlocks::Bool, resources::Nothing, debug::Bool, suppressWarnings::Bool, pkg_comp_kwargs::@Kwargs{})
    @ FMIBuild C:\Users\Owner\.julia\packages\FMIBuild\zfhlh\src\FMIBuild.jl:304
 [14] saveFMU
    @ C:\Users\Owner\.julia\packages\FMIBuild\zfhlh\src\FMIBuild.jl:72 [inlined]
 [15] top-level scope
    @ C:\Users\Owner\Documents\myJulia\Test\src\BouncingBall.jl:143
 [16] include(fname::String)
    @ Base.MainInclude .\client.jl:489
 [17] top-level scope
    @ REPL[10]:1
in expression starting at C:\Users\Owner\Documents\myJulia\Test\src\BouncingBall.jl:143

```

Also, BouncingBall.jl:143 means  this line that is the last part.
`saveFMU(fmu, fmu_save_path; compress=false) `

Does this error ring a bell?

**Comment by basatsu** [comment]
As a result, I just  want to save the model that I created as below as FMU file by FMIExport.jl.
This is what I expecting, but I can't figure out how to do it.
It'd be nice if you could give me your advice.

```
using Flux
using Statistics
using BSON: @save
# create random data here
input_size = 7
output_size = 2
num_samples = 1000
X = rand(Float32, input_size, num_samples)
Y = rand(Float32, output_size, num_samples)
# define model
model = Chain(
    Dense(input_size, 32, relu),
    Dense(32, 16, relu),
    Dense(16, output_size)
)
# define Loss Functions and Optimizers
loss(x, y) = Flux.mse(model(x), y)
opt = ADAM(0.01)
# training
epochs = 100
for epoch in 1:epochs
    Flux.train!(loss, Flux.params(model), [(X, Y)], opt)
    if epoch % 10 == 0
        @show epoch
        @show loss(X, Y)
    end
end
# save as BSON file
@save "vehicle_model.bson" model
```

**Comment by ThummeTo** [comment]
I just executed the current bouncing ball example under Julia 1.10.5 (Windows) with `saveFMU(fmu, fmu_save_path; compress=false)`.

I can't reproduce your error, compilation runs fine for me. It is important to not only copy the file, but also the package around! Important: FMIExport compiles packages, not only the file. Also check, that naming is consistent, so it is assumed, that file, package, project.toml and the project name within etc. have the same name (if you like to rename it).

If you copy the complete bouncing ball package (so the folder) and execute the file within (you can change `debug` if you want), it should work.

**Reply by basatsu** [reply]
Much appreciated. I'll try that.

**Reply by basatsu** [reply]
I end up getting the error....

Here is what I did. Also my Julia version is the latest 1.11.1 .


I created a folder named "BouncingBall".
Then 
`julia> activate BouncingBall`


`(BouncingBall) pkg> add FMIExport FMIBase FMICore`

```
(BouncingBall) pkg> status
Project BouncingBall v0.1.0
Status `C:\Users\Owner\Documents\myJulia\BouncingBall\Project.toml`
  [0c46a032] DifferentialEquations v7.14.0
  [900ee838] FMIBase v1.0.10
  [226f0e26] FMIBuild v0.3.2
  [8af89139] FMICore v1.1.1
  [31b88311] FMIExport v0.4.0
  [7073ff75] IJulia v1.25.0
```

I createe BouncingBall.ipynb in "src" folder then paste  your "BouncingBall" example code.
However this error is what happens when I try to execute it.
```
AssertionError: ["fmiBuild(...): Cannot find a package where this file is stored in. For FMU-Export, this source file needs to be inside of a package."]

Stacktrace:
 [1] saveFMU(fmu::FMIBase.FMU2, fmu_path::String, fmu_src_file::Nothing; standalone::Bool, compress::Bool, cleanup::Bool, removeLibDependency::Bool, removeNoExportBlocks::Bool, resources::Nothing, debug::Bool, suppressWarnings::Bool, pkg_comp_kwargs::@Kwargs{})
   @ FMIBuild C:\Users\Owner\.julia\packages\FMIBuild\zfhlh\src\FMIBuild.jl:184
 [2] saveFMU
   @ C:\Users\Owner\.julia\packages\FMIBuild\zfhlh\src\FMIBuild.jl:72 [inlined]
 [3] top-level scope
   @ In[10]:1
```
<img width="873" alt="image" src="https://github.com/user-attachments/assets/ba54c152-61c2-45eb-85ec-68dfa72a2470">

I'm executeing on Jupyter Notebook.
Where am I missing? Why do I get this error?

**Reply by basatsu** [reply]
Something is off....
Also I executed BouncingBall.jl (not only jupyter notebook file but also julia script file) with
`include("BouncingBall.jl")`

But I end up getting another error like this.

```
ERROR: The following 1 direct dependency failed to precompile:

BouncingBall

Failed to precompile BouncingBall [132c8a07-7570-42bd-ae26-6dbda4580768] to "C:\\Users\\Owner\\.julia\\compiled\\v1.11\\BouncingBall\\jl_9439.tmp".
[ Info: Saving example files at: C:\Users\Owner\AppData\Local\Temp\fmibuildjl_test_0zBhoQ
ERROR: LoadError: UndefVarError: `saveFMU` not defined in `BouncingBall`
Stacktrace:
 [1] top-level scope
```

Here is the code I tried.  This is the same  to your example code.

I just brought only `using FMIBuild:saveFMU` to the top part as you see below,though....

I'm not quite sure....

```
# module BouncingBall
# greet() = print("Hello World!")
#end # module BouncingBall

using FMIExport
using FMIExport.FMIBase.FMICore: fmi2True, fmi2False
using FMIBuild:saveFMU #bring saveFMU definition here

EPS = 1e-6

FMU_FCT_INIT = function ()
    m = 1.0        
    r = 0.0         
    d = 0.7         
    v_min = 1e-1    
    g = 9.81       
    sticking = fmi2False

    s = 1.0        
    v = 0.0       
    a = 0.0        

    t = 0.0
    x_c = [s, v]
    ẋ_c = [v, a]
    x_d = [sticking]
    u = []
    p = [m, r, d, v_min, g]

    return (t, x_c, ẋ_c, x_d, u, p)
end

FMU_FCT_EVALUATE = function (t, x_c, ẋ_c, x_d, u, p, eventMode)
    m, r, d, v_min, g = p
    s, v = x_c
    sticking = x_d[1]
    _, a = ẋ_c

    if sticking == fmi2True
        a = 0.0
    elseif sticking == fmi2False
        if eventMode
            if s < r && v < 0.0
                s = r + EPS 
                v = -v * d

              
                if abs(v) < v_min
                    sticking = fmi2True
                    v = 0.0
                end
            end
        else
          
        end

        a = (m * -g) / m   
    else
        @error "Unknown value for `sticking` == $(sticking)."
        return (x_c, ẋ_c, x_d, p)
    end

    x_c = [s, v]
    ẋ_c = [v, a]
    x_d = [sticking]
    p = [m, r, d, v_min, g]

    return (x_c, ẋ_c, x_d, p)
end

FMU_FCT_OUTPUT = function (t, x_c, ẋ_c, x_d, u, p)
    m, r, d, v_min, g = p
    s, v = x_c
    _, a = ẋ_c
    sticking = x_d[1]

    y = [s]

    return y
end

FMU_FCT_EVENT = function (t, x_c, ẋ_c, x_d, u, p)
    m, r, d, v_min, g = p
    s, v = x_c
    _, a = ẋ_c
    sticking = x_d[1]

    if sticking == fmi2True
        z1 = 1.0           
    else
        z1 = (s - r)         
    end

    z = [z1]

    return z
end


FMIBUILD_CONSTRUCTOR = function (resPath="")
    fmu = fmi2CreateSimple(initializationFct=FMU_FCT_INIT,
        evaluationFct=FMU_FCT_EVALUATE,
        outputFct=FMU_FCT_OUTPUT,
        eventFct=FMU_FCT_EVENT)

    fmu.modelDescription.modelName = "BouncingBall"


    fmi2ModelDescriptionAddModelExchange(fmu.modelDescription, "BouncingBall")


    fmi2AddStateAndDerivative(fmu, "ball.s"; stateDescr="Absolute position of ball center of mass", derivativeDescr="Absolute velocity of ball center of mass")
    fmi2AddStateAndDerivative(fmu, "ball.v"; stateDescr="Absolute velocity of ball center of mass", derivativeDescr="Absolute acceleration of ball center of mass")


    fmi2AddIntegerDiscreteState(fmu, "sticking"; description="Indicator (boolean) if the mass is sticking on the ground, as soon as abs(v) < v_min")


    fmi2AddRealOutput(fmu, "ball.s_out"; description="Absolute position of ball center of mass")


    fmi2AddRealParameter(fmu, "m"; description="Mass of ball")
    fmi2AddRealParameter(fmu, "r"; description="Radius of ball")
    fmi2AddRealParameter(fmu, "d"; description="Collision damping constant (velocity fraction after hitting the ground)")
    fmi2AddRealParameter(fmu, "v_min"; description="Minimal ball velocity to enter on-ground-state")
    fmi2AddRealParameter(fmu, "g"; description="Gravity constant")

    fmi2AddEventIndicator(fmu)

    return fmu
end


tmpDir = mktempdir(; prefix="fmibuildjl_test_", cleanup=false)
@info "Saving example files at: $(tmpDir)"
fmu_save_path = joinpath(tmpDir, "BouncingBall.fmu")

fmu = FMIBUILD_CONSTRUCTOR()

saveFMU(fmu, fmu_save_path; compress=false)


```

**Reply by ThummeTo** [reply]
I would recommend making a copy of the bouncing ball example folder to start with, instead of creating a new package from scratch.

Also note, that there are differences between the original example and your code, e.g. it is important to place the `saveFMU` call inside a `### FMIBUILD_NO_EXPORT_BEGIN ###` block. Even if this is only a comment, this code is removed during export (see the comment above this line in the example), which is important and what you want - otherwise you would export an FMU that exports FMUs (infinite loop) :-)

I also recommend to not use a jupyter notebook in this case, because it's much more complicated to debug and you can't terminate code execution easily.

**Comment by ThummeTo** [comment]
(I convert this to discussion because it doesn't seem to be a bug)

**Reply by basatsu** [reply]
@ThummeTo 
Sorry to bother you.

By the way, is it possible to  export the model that I created with Flux.jl as my above post to FMU file?
This is what I want as a result and I'm doing trial and error but if possible, I wonder how we should do it with FMIExport.jl and stuff....
Do you have any ideas?

**Reply by ThummeTo** [reply]
No problem.

Check out the "neural FMU" example, there we also include some neural networks. Thats a good starting point.

In theory (and I think in practice too), this should work (I did a test a year ago, exporting a Flux model). The big problem is, that Flux is a heavy dependency and compiling Flux including its dependencies to a FMU is not what you want. In general you want only the NN inference. That's why we coded this in the "neural FMU" by ourself (if you look close, you see custom functions for NN inference, so we don't need to include Flux).

**Reply by basatsu** [reply]
I checked out "neural FMU" example, but do we need to use load an existing FMU?
```
# loads an existing FMU
fmu = loadFMU(joinpath(resPath, "SpringDamperPendulum1D.fmu"))

```

I mean, do we have to use FMIImport?

In my case, I feel like I don't need "load FMU" and I will create NN model from scratch...
I'm having difficulty understanding "neural FMU" example....
I wonder how the example works and where I can replace to my above code.

**Reply by ThummeTo** [reply]
If you want to have a FMU as part of your new (to be exported) FMU, you need `loadFMU`. If you only want an ANN without anything else, you don't need `loadFMU` of course.

**Reply by ThummeTo** [reply]
However, this example exports further an ANN, which is a good starting point for you I think.
